src/Microsoft.Azure.WebJobs.Host/Hosting/FunctionDataCache/ICacheAwareReadObject.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public interface ICacheAwareReadObject : IDisposable
+    {
+        /// <summary>
+        /// Gets or sets if the object was a cache hit.
+        /// If this is <see cref="true"/> (i.e. it was a cache it) then the <see cref="CacheObject"/> property must be used to
+        /// read the object from the cache.
+        /// Otherwise, the <see cref="BlobStream"/> property must be used to read it from storage.
+        /// </summary>
+        bool IsCacheHit { get; }
+
+        /// <summary>
+        /// Gets or sets the key corresponding to the object pointed to by this Stream.
+        /// </summary>
+        FunctionDataCacheKey CacheKey { get; }
+
+        /// <summary>
+        /// Gets or sets the metadata about the shared memory region containing content of the object pointed to by this Stream.
+        /// </summary>
+        SharedMemoryMetadata CacheObject { get; }
+
+        /// <summary>
+        /// Gets or sets the Stream pointing to this object in storage.
+        /// </summary>
+        Stream BlobStream { get; }
+
+        /// <summary>
+        /// Put the object into the cache.
+        /// </summary>
+        /// <param name="cacheObject"></param>
+        /// <param name="isIncrementActiveReference"></param>
+        /// <returns><see cref="true"/> if the object was successfully put into the cache, <see cref="false"/> otherwise.</returns>
+        bool TryPutToCache(SharedMemoryMetadata cacheObject, bool isIncrementActiveReference);
```

Comment:
I am confused why the read object is writing to cache?
Also unsure what isIncrementActiveReference does.

Reply:
There are two cases where an object is inserted into the cache:
1) An output is produced (i.e., `ICacheAwareWriteObject`) and we want to persist it into the cache for potential future reads.
2) Ann input is read from storage (i.e., `ICacheAwareReadObject`) and we want to persist it into the cache for potential re-reads by subsequent function invocations.

I will add a docstring for `isIncrementActiveReference`.

Reply:
technically you could use either object for both purposes. In that case, should this be merged into a single type. CacheAwareObject